File: poseidon/examples/PoseidonNative.kt
package xyz.atoshi.poseidon

import java.math.BigInteger

/**
 * Native JNI bridge to libatoshi_poseidon.so.
 *
 * Setup:
 *   1. Run ../build-android.sh — outputs jniLibs/<abi>/libatoshi_poseidon.so
 *   2. Copy jniLibs/ into your Android project: app/src/main/jniLibs/
 *   3. Place this file in app/src/main/java/xyz/atoshi/poseidon/PoseidonNative.kt
 *      (the package path xyz.atoshi.poseidon is hard-coded in the JNI symbol name)
 *
 * Usage:
 *   val commitment = PoseidonNative.hash(amount, tokenId, owner, blinding)
 *   val nullifier  = PoseidonNative.hash(commitment, privateKey, leafIndex)
 */
object PoseidonNative {
    init {
        System.loadLibrary("atoshi_poseidon")
    }

    /**
     * Low-level API. Compute Poseidon hash over N inputs (1..=16).
     *
     * @param inputs Array of 32-byte big-endian field elements.
     * @return 32-byte big-endian hash output.
     */
    @JvmStatic
    external fun hash(inputs: Array<ByteArray>): ByteArray

    /**
     * High-level API for ergonomic Kotlin use.
     * Pass decimal big-int strings (or "0x..." hex strings), returns decimal string.
     *
     * Usage:
     *   val out = PoseidonNative.hashStrings(arrayOf(
     *       amount.toString(),      // "1000000000000000000"
     *       tokenId.toString(),     // "0"
     *       owner.toString(),       // "12345..."
     *       blinding.toString(),    // "67890..."
     *   ))
     *   val commitment = BigInteger(out)
     */
    @JvmStatic
    external fun hashStrings(inputs: Array<String>): String

    /**
     * Convenience: take BigInteger inputs (must be < BN254 scalar field size),
     * encode them to 32-byte BE buffers, hash, and return the result as BigInteger.
     * Internally uses the ByteArray API.
     */
    fun hash(vararg fields: BigInteger): BigInteger {
        require(fields.isNotEmpty() && fields.size <= 16) {
            "Poseidon arity must be 1..=16, got ${fields.size}"
        }
        val bufs = Array(fields.size) { i -> toBe32(fields[i]) }
        val out = hash(bufs)
        return BigInteger(1, out)
    }

    /** Big-endian, zero-padded to 32 bytes. */
    private fun toBe32(v: BigInteger): ByteArray {
        require(v.signum() >= 0) { "negative not allowed" }
        val raw = v.toByteArray()
        // raw may have a leading 0x00 sign byte or be shorter than 32 bytes
        val out = ByteArray(32)
        val start = if (raw.size > 32) raw.size - 32 else 0
        val len = minOf(raw.size, 32)
        System.arraycopy(raw, start, out, 32 - len, len)
        return out
    }
}

File: poseidon/examples/PoseidonNative.swift
import Foundation

/// Native bridge to AtoshiPoseidon.xcframework (libatoshi_poseidon.a).
///
/// Setup:
///   1. Run ../build-ios.sh — outputs AtoshiPoseidon.xcframework + include/atoshi_poseidon.h
///   2. Drag AtoshiPoseidon.xcframework into Xcode (Embed & Sign).
///   3. Add include/atoshi_poseidon.h to a Swift bridging header:
///         #import "atoshi_poseidon.h"
///
/// Usage:
///   let commitment = try PoseidonNative.hash([amount, tokenId, owner, blinding])
///   let nullifier  = try PoseidonNative.hash([commitment, privateKey, leafIndex])
public enum PoseidonError: Error {
    case invalidArity
    case hashFailed(Int32)
}

public enum PoseidonNative {
    /// Compute Poseidon hash over N 32-byte field elements (1..=16).
    public static func hash(_ inputs: [Data]) throws -> Data {
        guard !inputs.isEmpty, inputs.count <= 16 else {
            throw PoseidonError.invalidArity
        }
        for buf in inputs where buf.count != 32 {
            throw PoseidonError.invalidArity
        }

        var packed = Data(capacity: 32 * inputs.count)
        for buf in inputs { packed.append(buf) }

        var out = Data(count: 32)
        let status = packed.withUnsafeBytes { (inPtr: UnsafeRawBufferPointer) -> Int32 in
            out.withUnsafeMutableBytes { (outPtr: UnsafeMutableRawBufferPointer) -> Int32 in
                atoshi_poseidon_hash(
                    inPtr.bindMemory(to: UInt8.self).baseAddress,
                    inputs.count,
                    outPtr.bindMemory(to: UInt8.self).baseAddress
                )
            }
        }
        guard status == 0 else {
            throw PoseidonError.hashFailed(status)
        }
        return out
    }
}
